### PR TITLE
Add MTE-1520 [v119] Add simple test for Fakespot

### PR DESCRIPTION
**Pull request by clarmso**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/MTE-1520)

## :bulb: Description
Fakespot has been enabled on Amazon.com. Let me write a new test to ensure that Fakespot is present.

## :pencil: Checklist
You have to check all boxes before merging
- [X] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

File: Tests/XCUITests/FakespotTests.swift
Comment (dragosb01) [inline Tests/XCUITests/FakespotTests.swift:7]: Please add the testrail link before the test

File: Tests/XCUITests/FakespotTests.swift
Comment (thatswinnie) [inline Tests/XCUITests/FakespotTests.swift:22]: Could we use the accessibility ids to look for view elements? The id for the button is `AccessibilityIdentifiers.Toolbar.shoppingButton`.

File: Tests/XCUITests/FakespotTests.swift
Comment (thatswinnie) [inline Tests/XCUITests/FakespotTests.swift:25]: The title has the a11y id `AccessibilityIdentifiers.Shopping.sheetHeaderTitle`.

File: Tests/XCUITests/FakespotTests.swift
Comment (isabelrios) [reply Tests/XCUITests/FakespotTests.swift:7]: If not created yet, we should create the tests there, I already talked to @abodea about this and he agrees on having the test suite there with this basic test for now

**Comment by isabelrios** [comment]
I run the test locally and it works consistently. Good job here! 
I agree that the best place to start with is running it on the Full Functional, we can move it to block PRs on the smoketest when we feel confident about it and the feature not changing.

File: Tests/XCUITests/FakespotTests.swift
Comment (clarmso) [reply Tests/XCUITests/FakespotTests.swift:22]: The `app.debugDescription` does not show `AccessibilityIdentifiers.Toolbar.shoppingButton` in the awesome bar. Such an icon is displayed and can be tapped. Here is the debugDescription:
```
                        Other, 0x7ff510a1bb60, {{7.0, 53.0}, {376.0, 42.0}}
                          TextField, 0x7ff510a3e0d0, {{48.0, 63.0}, {215.0, 22.0}}, identifier: 'url', label: 'Address Bar', placeholderValue: 'Search or enter address', value: www.amazon.com/gp/...
                          Button, 0x7ff510a3a870, {{343.0, 54.0}, {40.0, 40.0}}, identifier: 'TabLocationView.reloadButton', label: 'Reload page'
                          Button, 0x7ff510a3ac20, {{7.0, 54.0}, {40.0, 40.0}}, identifier: 'TabLocationView.trackingProtectionButton', label: 'Secure connection'
                          Button, 0x7ff510a3afd0, {{303.0, 54.0}, {40.0, 40.0}}, identifier: 'TabLocationView.shareButton', label: 'Share this page'
  ```

Comment (isabelrios) [reply Tests/XCUITests/FakespotTests.swift:22]: @thatswinnie has that icon change the ID, wondering why @clarmso cannot see it when printint the elements... I will give this a try later today

Comment (isabelrios) [reply Tests/XCUITests/FakespotTests.swift:22]: I can't see it either... only Tracking Protection, Share and Reload button are available via print(app.debugDescription)
Screenshot 2023-09-18 at 11 41 05

Comment (clarmso) [reply Tests/XCUITests/FakespotTests.swift:22]: Relying on the coordinates to tap on the `shoppingButton` (via `tapAtPoint`) is flaky on my end. I would not merge this PR before `debugDescription` could see the icon. :sa

File: Tests/XCUITests/FakespotTests.swift
Comment (clarmso) [reply Tests/XCUITests/FakespotTests.swift:8]: I have created the test on TestRail.

File: Tests/XCUITests/FakespotTests.swift
Comment (clarmso) [inline Tests/XCUITests/FakespotTests.swift:25]: Using the coordinates to tap the shopping icon is flaky.

File: Tests/XCUITests/FakespotTests.swift
Comment (isabelrios) [reply Tests/XCUITests/FakespotTests.swift]: Agreed... that coordinates are specific to a device's screen size.. better to have the ID...

**Comment by clarmso** [comment]
The test fails intermittently because Amazon may issue a Captcha-like challenge when the site is hit repeatedly. If we don't run the test repeatedly, the intermittent failures are avoidable.

Screenshot 2023-09-26 at 3 53 58 PM
Screenshot 2023-09-26 at 3 54 08 PM

File: Tests/XCUITests/FakespotTests.swift
Comment (thatswinnie) [inline Tests/XCUITests/FakespotTests.swift:27]: 🎉

File: Tests/XCUITests/FakespotTests.swift
Comment (thatswinnie) [inline Tests/XCUITests/FakespotTests.swift:30]:
```suggestion
            XCTAssertEqual(app.staticTexts[AccessibilityIdentifiers.Shopping.sheetHeaderTitle].label, String.Shopping.SheetHeaderTitle)
```
This requires to add `import Shared` as well

File: Tests/XCUITests/FakespotTests.swift
Comment (thatswinnie) [inline Tests/XCUITests/FakespotTests.swift:33]: This is an example where the devs should have been notified because the a11y id is missing for this element. I added it in this [PR](https://github.com/mozilla-mobile/firefox-ios/pull/16609). Once merged this would be a better approach:
```suggestion
            app.otherElements.buttons[AccessibilityIdentifiers.Shopping.sheetCloseButton].tap()
```

File: bitrise.yml
Comment (isabelrios) [inline bitrise.yml:105]: Just curious why do we need this?

**Comment by isabelrios** [comment]
This is looking great @clarmso! Once the IDs are added as @thatswinnie commented, this could be ready to land.
I would not worry for the # of test executions for now because this will run as part of the Full Functional. That said, I think we need to think of a way, if possible, to avoid that issue, let's explore which options we would have.
Thanks both for your work with this PR! 🙏

**Review by thatswinnie — APPROVED** [review]

**Comment by isabelrios** [comment]
Bitrise is red due to the increase in warnings. There is a new warning added in this PR: 
```
/Users/vagrant/git/Tests/XCUITests/FakespotTests.swift:5:8: implicit import of bridging header 'Shared-Bridging-Header.h' via module 'Shared' is deprecated and will be removed in a later version of Swift
import Shared
```

File: bitrise.yml
Comment (clarmso) [reply bitrise.yml]: I was trying to debug the lint warning/error.

**Comment by clarmso** [comment]
> This is looking great @clarmso! Once the IDs are added as @thatswinnie commented, this could be ready to land. I would not worry for the # of test executions for now because this will run as part of the Full Functional. That said, I think we need to think of a way, if possible, to avoid that issue, let's explore which options we would have. Thanks both for your work with this PR! 🙏

@isabelrios The test fails about 5-6 times out of 100 consecutive test runs. If we only run the test once every few minutes, we may be ok. 🤔

File: Tests/XCUITests/FakespotTests.swift
Comment (clarmso) [inline Tests/XCUITests/FakespotTests.swift:30]: @thatswinnie `String.Shopping.SheetHeaderTitle` is not recognized. Please help. 😞

Comment (isabelrios) [reply Tests/XCUITests/FakespotTests.swift:30]: looks like we need `import Shared` for that to be recognized but then there is the new warning...

File: Tests/XCUITests/FakespotTests.swift
Comment (clarmso) [inline Tests/XCUITests/FakespotTests.swift:32]: I followed the same style as in the source code here.
https://github.com/mozilla-mobile/firefox-ios/blob/e44e65bfcbb015264db5b64ea01863d4328a8be8/Client/Frontend/Fakespot/FakespotViewController.swift#L47

File: Tests/XCUITests/FakespotTests.swift
Comment (clarmso) [inline Tests/XCUITests/FakespotTests.swift:5]: `import Shared` is giving me a new warning but the identifier does not work without this import. 🤔 
```
⚠️  /Users/vagrant/git/Tests/XCUITests/FakespotTests.swift:5:8: implicit import of bridging header 'Shared-Bridging-Header.h' via module 'Shared' is deprecated and will be removed in a later version of Swift
import Shared
```

Comment (isabelrios) [reply Tests/XCUITests/FakespotTests.swift:5]: @clarmso we should try to fix this warning before merging, otherwise we will be close to threshold again...

**Comment by isabelrios** [comment]
@clarmso please rebase latest main, unit tests failures should be gone with that and we could land this PR

**Review by isabelrios — APPROVED** [review]

**Pull request merged**
